src/pso.c

```diff
@@ -0,0 +1,204 @@
+      size_t idx = (particle*dim) + d;
+      vnew[idx] = INERTIA*v[idx] + COG*random_0_to_1()*(y[idx] - x[idx]) +
+      SOCIAL*random_0_to_1()*(best[d] * x[idx]);
+      vnew[idx] = min(max(min_vel[d],vnew[idx]),max_vel[d]);
+    }
+  }
+  return vnew;
+}
+
+void update_position(double* x, double* v,
+                    size_t swarm_size, size_t dim,
+                    const double* const min_positions,
+                    const double* const max_positions) {
+  for (size_t particle=0; particle < swarm_size; particle++){
+    for (size_t d=0; d<dim; d++){
+      size_t idx = (particle*dim) + d;
+      x[idx] += v[idx];
+      if (x[idx] >= max_positions[d] || x[idx] <= min_positions[d]){
+        v[idx] = -v[idx];
+      }
+      x[idx] = min(max(min_positions[d],x[idx]),max_positions[d]);
+      //TODO: PROJECT ON BOUNDS - RATHER DO A BOUNCE THAN A SIMPLE PROJECT
+    }
+  }
+  return;
+}
+
+  /**
+  Particle Swarm Algorithm to find the global minima of objective function using a nature mimicking swarm
+  **/
+double * pso_basic (double(*fit)(const double* const, size_t),
+                    size_t swarm_size,
+                    size_t dim,
+                    size_t max_iter,
+                    const double* const min_positions,
+                    const double* const max_positions) {
+  srand((unsigned) time(NULL));
+  // REPLACE WITH FIXED ZERO SEED
+
+  // double known_optima = 0.;   // TODO : DEPENDENT ON OBJECTIVE FUNCTION
+  double* min_vel = (double*)malloc(dim*sizeof(double));
+  double* max_vel = (double*)malloc(dim*sizeof(double));
+  generate_vel_limit(min_positions,max_positions,min_vel,max_vel,dim);
+
+  size_t sizeof_position = dim*swarm_size*sizeof(double);
+  // randomly initialise postions of swarm particles
+  double* current_positions = (double*)malloc(sizeof_position);
+  current_positions = rand_init(swarm_size, dim,min_positions,max_positions);
+  double* local_best_position = (double*)malloc(sizeof_position);
+  memcpy(local_best_position,current_positions, sizeof_position);
+
+  // calculate the fitness at every position of the swarm
+  size_t sizeof_fitness = swarm_size*sizeof(double);
+  double* current_fitness = (double*)malloc(sizeof_fitness);
+  eval_fitness(fit,swarm_size,dim,current_positions,current_fitness);
+  double* local_best_fitness = (double*)malloc(sizeof_fitness);
+  memcpy(local_best_fitness,current_fitness, sizeof_fitness );
+
+  double* p_velocity = (double*)malloc(sizeof_position);
+  p_velocity = gen_init_velocity(current_positions,swarm_size,dim,
+  min_positions,max_positions);
+
+  double* global_best_position = (double*)malloc(dim*sizeof(double));
+  size_t global_best_idx = best_fitness(local_best_fitness,dim,swarm_size);
+
+  global_best_position = (local_best_position + (dim*global_best_idx));
+  double global_best_fitness = local_best_fitness[global_best_idx];
+
+  size_t iter = 0;
+  // double error = 1e3;
+  while ( iter < max_iter) { // error > EPS &&

```

Comment:
Should this code be in the condition or not?

Reply:
Yeah ideally if we were trying to use the function to optimize, but for benchmarking, having the while loop dependent on the error is not great, since we cant ensure that a certain number of  iterations take place. 
Hence it's commented out. 
I'm  removing it actually, since whatever info is needed can be achieved from the integration tests.

Reply:
Makes sense